larq_compute_engine/mlir/transforms/prepare_patterns_common.td

```diff
@@ -4,9 +4,56 @@ include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 include "larq_compute_engine/mlir/ir/lce_ops.td"
 include "larq_compute_engine/mlir/transforms/op_removal_patterns.td"
 
+class ConstantValue<string val> : AttrConstraint<CPred<"IsConstantValue($_self, " # val # ")">>;
 
-// This relies on implementation details of larq.math.sign. We should make
-// this more general in the future
+def CreateTFBroadcastToOp : NativeCodeCall<
+    "$_builder.create<TF::BroadcastToOp>("
+        "$0.getLoc(),"
+        "RankedTensorType::get("
+            "$0.getType().cast<RankedTensorType>().getShape(),"
+            "getElementTypeOrSelf($1.getType())),"
+        "$1,"
+        "$2)">;
+
+def CreateTFShapeOp : NativeCodeCall<
+    "$_builder.create<TF::ShapeOp>($0.getLoc(), $1, $2)">;
```

Comment:
To make sure I understand this correctly: Why does the `tf.shape` op take two input arguments instead of just one? Or am I missing something here?

Reply:
This is the definition as [used in TF here](https://github.com/tensorflow/tensorflow/blob/25a9edf48c4a26ee3b78eaec72cec6281c44ead7/tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td#L43-L44).

It takes three args — `$0` is used to get the correct location, `$1` is the input (thing we're getting the shape of), and `$2` is an attribute for the shape op that is annotated "use 32bit" (see e.g. [TF usage here](https://github.com/tensorflow/tensorflow/blob/25a9edf48c4a26ee3b78eaec72cec6281c44ead7/tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td#L419-L420)).

Reply:
Ahh I see, thanks that makes sense.